File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
//         if(root == null)
//         {
//             return true;
//         }
//         if(root.left==null || root.right==null)
//         {
//             return true;
//         }
//         if(root.left.val< root.val && root.right.val>root.val)
//         {
//             return true;
//         }
//         else
//         {
//             return false;
//         }
        
        
//         return isValidBST(root.left) && isValidBST(root.right);
        return validate(root,Double.NEGATIVE_INFINITY,Double.POSITIVE_INFINITY);
    }
    private boolean validate(TreeNode root, double left, double right )
    {
        if(root==null)
        {
            return true;
        }
        if(!(left<root.val && right>root.val))
        {
            return false;
        }
        return validate(root.left,left,root.val) && validate(root.right, root.val, right);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(curr, minVal, maxVal):
            if not curr:
                return True;
            if not (minVal<curr.val<maxVal):
                return False;
            
            return dfs(curr.left,minVal,curr.val) and dfs(curr.right,curr.val,maxVal);
        
        return dfs(root,float(-inf),float(inf));
        
            
        